Use boolean not bool for libjpeg functions

ryandesign opened the pull request:
Use boolean `TRUE` instead of bool `true` for libjpeg functions. Fixes build failure:

```
../../../../../src/emu/rendutil.cpp:580:2: error: no matching function for call to 'jpeg_read_header'
        jpeg_read_header(&cinfo, true);
        ^~~~~~~~~~~~~~~~
In file included from ../../../../../src/emu/rendutil.cpp:16:
/opt/local/include/jpeglib.h:1039:13: note: candidate function not viable: no known conversion from 'bool' to 'boolean' for 2nd argument
EXTERN(int) jpeg_read_header JPP((j_decompress_ptr cinfo,
            ^
```

cuavas commented:
Which OS/toolchain/runtime library (including versions) are you building with when you get this error?  Are you using any unusual compile options?

ryandesign commented:
macOS High Sierra, Apple LLVM version 9.1.0 (clang-902.0.39.2) from Xcode 9.4.1, not sure what you mean by runtime library, but I am using libjpeg 9c from MacPorts. No unusual compile options that I know of.

This is a pretty common problem. I discussed it with a developer of libjpeg a few years ago when we saw various projects fail to build after I updated MacPorts libjpeg to version 9 and removed a patch we had previously been using to change the behavior of booleans. See specifically his response to our observed build failure of digikam, which was also using `true` instead of the required `TRUE`: "The libjpeg API specifies the type 'boolean' here with values TRUE or FALSE, and everything else is an error and always was":

> Begin forwarded message:
> 
> From: Guido Vollbeding <guido@jpegclub...>
> Subject: Re: Build failures with jpeg 9a
> Date: March 6, 2014 at 07:07:17 CST
> To: Ryan Schmidt <ryandesign@macports...>
> 
> Dear Ryan
> 
> Thank you for feedback.
> 
> In version 9 we changed to a more reliable definition of the 'boolean'
> type (in file jmorecfg.h).
> This may cause conflicts with applications which do not comply with the
> specified application programming interface (the digikam issue is such
> a case).
> In other cases, it may be necessary to use the HAVE_BOOLEAN mechanism,
> if application insists on using their own type.
> We have improved the situation in 9a, but some cases remain which can
> only be fixed in application code.
> 
> Our suggestion is that software developers be notified to correct or
> adapt their programs appropriately.
> I have done this with the libjpeg maintainer of the Debian Linux
> distribution, in which course we went through a batch of cases which
> were affected by the issue, and I offered respective solutions.
> Your cases were part of the batch:
> 
> digikam:
> If "true" is a variable of int type, then using an explicit typecast
> "(boolean)true" fixes the issue.
> If "true" means a constant of some other bool type, then using TRUE
> instead fixes the issue.
> The libjpeg API specifies the type "boolean" here with values TRUE
> or FALSE, and everything else is an error and always was.
> 
> swftools-0.9.2 defines following:
> 
>  lib/jpeg.c:
> 
>  #ifdef HAVE_JPEGLIB
>  #define HAVE_BOOLEAN
>  #include <jpeglib.h>
> 
> The interpretation of HAVE_BOOLEAN has been changed in new library,
> which now means that both the type AND the corresponding values are
> defined, which is more general and consistent.
> Modifying the place where boolean is defined or changing at this place to
> 
>  #ifdef HAVE_JPEGLIB
>  #ifndef FALSE            /* in case these macros already exist */
>  #define FALSE   0        /* values of boolean */
>  #endif
>  #ifndef TRUE
>  #define TRUE    1
>  #endif
>  #define HAVE_BOOLEAN
>  #include <jpeglib.h>
> 
> should make it work with either version of libjpeg.
> 
> Any other issues should be variations of these and could be resolved
> similarly.  If you have particular problem please let me know and I
> will look at it.
> 
> Regards
> Guido Vollbeding
> Organizer Independent JPEG Group

ryandesign commented:
BTW we have been using the patch in this PR ever since we added mame to MacPorts last year; see https://trac.macports.org/ticket/54699#comment:6.

cuavas commented:
So you're not building with the bundled libjpeg in 3rdparty/libjpeg - that's not a first-class supported configuration, and MAME isn't tested or guaranteed to work with other versions of libjpeg.  Having common words like `boolean`, `TRUE` and `FALSE` leak from libjpeg public headers is really nasty and things will potentially break if something else provides definitions that differ from what got used when the library was compiled.

That said, the bundled version of libjpeg does define `TRUE`/`FALSE`, so this patch is probably safe.  But after looking at the kinds of things libjpeg headers define, it's probably a good idea at some point to wrap the code that depends on libjpeg somewhere in `src/lib/util` so the macros don't leak out into other code.
